llrt_modules/src/modules/path.rs

```diff
@@ -1,267 +1,292 @@
+        .unwrap_or_default();
+    let filename = path
+        .file_name()
+        .map(|n| n.to_str().unwrap())
+        .unwrap_or_default();
+
+    let (name, extension) = name_extname(filename);
+
+    let root = path
+        .components()
+        .next()
+        .and_then(|c| match c {
+            Component::Prefix(prefix) => prefix.as_os_str().to_str(),
+            Component::RootDir => c.as_os_str().to_str(),
+            _ => Some(""),
+        })
+        .unwrap_or_default();
+
+    obj.set("root", root)?;
+    obj.set("dir", parent)?;
+    obj.set("base", [name, extension].concat())?;
+    obj.set("ext", extension)?;
+    obj.set("name", name)?;
+
+    Ok(obj)
+}
+
+fn join(parts: Rest<String>) -> String {
+    join_path(parts.0.iter())
+}
+
+pub fn join_path<S, I>(parts: I) -> String
+where
+    S: AsRef<str>,
+    I: Iterator<Item = S>,
+{
+    let mut empty = true;
+    let result = parts.fold(PathBuf::new(), |mut acc, part| {
+        let part = part.as_ref();
+        if part.starts_with(SEP_PAT) && empty {
+            acc.push(MAIN_SEPARATOR_STR);
+            empty = false;
+        }
+        for sub_part in part.split(SEP_PAT) {
+            if !sub_part.is_empty() {
+                if sub_part.starts_with("..") {
+                    empty = false;
+                    acc.pop();
+                } else {
+                    let sub_part = sub_part.strip_prefix('.').unwrap_or(sub_part);
+                    acc.push(sub_part);
+                    empty = false;
+                    #[cfg(windows)]
+                    if sub_part.ends_with(":") {
+                        acc.push(MAIN_SEPARATOR_STR);
+                    }
+                }
+            }
+        }
+        acc
+    });
+
+    remove_trailing_slash(result)
+}
+
+fn remove_trailing_slash(result: PathBuf) -> String {
+    let path = result.to_string_lossy().to_string();
+    path.strip_suffix(SEP_PAT).unwrap_or(&path).to_string()
+}
+
+fn resolve(path: Rest<String>) -> String {
+    resolve_path(path.iter())
+}
+
+pub fn resolve_path<S, I>(parts: I) -> String
+where
+    S: AsRef<str>,
+    I: Iterator<Item = S>,
+{
+    let result = parts.fold(std::env::current_dir().unwrap(), |mut acc, part| {
+        let part = part.as_ref();
+        if is_absolute(part.into()) {
+            acc = PathBuf::from(part);
+        } else {
+            for sub_part in part.split(SEP_PAT) {
+                if sub_part.starts_with("..") {
+                    acc.pop();
+                } else if sub_part == "." {
+                    // skip
+                } else if sub_part.starts_with("./") {
+                    acc.push(sub_part.strip_prefix("./").unwrap_or(sub_part))
+                } else if sub_part.starts_with(".\\") {
+                    acc.push(sub_part.strip_prefix(".\\").unwrap_or(sub_part))
+                } else {
+                    acc.push(sub_part)
+                }
+            }
+        }
+        acc
+    });
+

```

Comment:
I think this method can be heavily optimized to something like this:

```rust
use memchr::memchr2;

fn resolve_path<S, I>(parts: I) -> PathBuf
where
    S: AsRef<str>,
    I: Iterator<Item = S>,
{
    parts.fold(std::env::current_dir().unwrap(), |mut acc, part| {
        let part = part.as_ref();
        if Path::new(part).is_absolute() {
            acc = PathBuf::from(part);
        } else {
            let mut start = 0;
            while start < part.len() {
                let end = find_next_separator(&part[start..]).map_or(part.len(), |i| i + start);
                match &part[start..end] {
                    ".." => { acc.pop(); }
                    "." => {}
                    sub_part => acc.push(sub_part),
                }
                start = end + 1;
            }
        }
        acc
    })
}

#[cfg(windows)]
fn find_next_separator(s: &str) -> Option<usize> {
    memchr2(b'\\', b'/', s.as_bytes())
}

#[cfg(not(windows))]
fn find_next_separator(s: &str) -> Option<usize> {
    s.find(std::path::MAIN_SEPARATOR)
}

// Constants kept for potential use elsewhere
#[cfg(windows)]
const SEP_PAT: [char; 2] = ['\\', '/'];
#[cfg(not(windows))]
const SEP_PAT: [char; 1] = [std::path::MAIN_SEPARATOR];
```

Instead of splitting and then checking prefixes and removing them I create a slice based on the next separator if present. memchr2 allows me to check of two instances of separators using inartistic instructions (and SIMD). Since we now have the slices we can push them to the path without needing to strip prefixes. A similar technique can be applied to `join_paths`